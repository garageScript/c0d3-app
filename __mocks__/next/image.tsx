import React from 'react'
import type { ImageProps } from 'next/image'
import _ from 'lodash'

const Image = (props: ImageProps) => {
  /*not all Image properties are allowed in <img/> tag
  ObjectFit causes warnings, but otherwise lessons/error images won't keep their dimensions 
  see https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit*/
  const copy = _.omit(props, 'objectFit')
  return <img {...copy} />
}

export default Image
