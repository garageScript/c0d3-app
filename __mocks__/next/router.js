const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
