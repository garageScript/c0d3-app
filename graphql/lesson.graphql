query Lesson {
  lessons {
    id
    title
  }
}