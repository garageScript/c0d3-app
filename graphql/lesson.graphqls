type Lesson {
  id: String
  title: String
}

type Query {
  lessons: Lesson!
}