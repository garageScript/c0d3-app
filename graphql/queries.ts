import { gql } from 'apollo-boost'

export const GET_LESSONS = gql`
  query Lessons {
    curriculumStatus {
      id
      description
      title
      order
      docUrl
      challenges {
        id
      }
      currentUser {
        userLesson {
          isEnrolled
          isTeaching
          isPassed
        }
      }
    }
    lessons {
      id
      title
      description
      order
      docUrl
      challenges {
        id
        description
        title
        order
      }
    }
  }
`

export const GET_LESSON = gql`
  query lessonInfo(
    $lessonInfo: LessonId
    $lessonUserInfo: LessonUserId
    $lessonStatus: LessonId
  ) {
    lessonInfo(input: $lessonInfo) {
      id
      title
      description
      order
      docUrl
      chatUrl
      challenges {
        id
        description
        title
        order
      }
    }
    userSubmissions(input: $lessonUserInfo) {
      id
      status
      mrUrl
      diff
      viewCount
      comment
      challengeId
      reviewer {
        id
        username
      }
      createdAt
      updatedAt
    }
    lessonStatus(input: $lessonStatus) {
      isPassed
      isTeaching
      isEnrolled
    }
  }
`

export const GET_SUBMISSIONS = gql`
  query submissions($in: LessonId) {
    submissions(input: $in) {
      status
    }
  }
`
