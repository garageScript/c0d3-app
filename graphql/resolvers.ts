import { login, logout, signup } from '../helpers/controllers/authController'
import db from '../helpers/dbload'

const { Lesson, User } = db

export default {
  Query: {
    lessons() {
      return Lesson.findAll({
        include: [
          'challenges',
          {
            model: User,
            through: {
              attributes: ['isPassed', 'isTeaching', 'isEnrolled']
            }
          }
        ],
        order: [['order', 'ASC']]
      })
    }
  },

  Mutation: {
    login,
    logout,
    signup
  }
}
