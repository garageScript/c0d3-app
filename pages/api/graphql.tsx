import { ApolloServer, gql } from 'apollo-server-micro'

const typeDefs = gql`
  type Query {
    username: String
    hello: String
  }
`

const resolvers = {
  Query: {
    username(_parent: any, _args: any, ctx: any) {
      return ctx.user
    },
    hello() {
      return 'hello'
    }
  }
}

const context = ({ req }: { req: any }) => req.cookies

const apolloServer = new ApolloServer({ typeDefs, resolvers, context })

export const config = {
  api: {
    bodyParser: false
  }
}

export default apolloServer.createHandler({ path: '/api/graphql' })
