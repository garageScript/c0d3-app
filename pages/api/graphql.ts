import { ApolloServer } from 'apollo-server-micro'
import session from 'express-session'
import nextConnect from 'next-connect'
import db from '../../helpers/dbload'
import typeDefs from '../../graphql/typeDefs'
import resolvers from '../../graphql/resolvers'
import connectSequelize from 'connect-session-sequelize'

// Sequelize Store for express-sessions
const { sequelize } = db
const SequelizeStore = connectSequelize(session.Store)

const ONE_WEEK = 1000 * 60 * 60 * 24 * 7

const handler: any = nextConnect() // For session middleware. TODO: Need to define types for Session
const apolloServer = new ApolloServer({
  typeDefs,
  resolvers,
  context: ({ req }) => ({ req }) // This lets GraphQL have access to sessions
})

const graphQLHandler = apolloServer.createHandler({ path: '/api/graphql' })

handler
  .use(
    session({
      secret: process.env.SESSION_SECRET || '',
      store: new SequelizeStore({
        db: sequelize
      }),
      resave: false,
      saveUninitialized: false,
      cookie: {
        domain: process.env.DB_HOST,
        maxAge: ONE_WEEK
      }
    })
  )
  .get('/api/graphql', graphQLHandler)
  .post('/api/graphql', graphQLHandler)

export const config = {
  api: {
    bodyParser: false // This is required for GraphQL to work properly  https://blog.logrocket.com/building-a-graphql-server-in-next-js/
  }
}

export default handler
