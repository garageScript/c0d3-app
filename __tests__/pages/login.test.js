import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GraphQLError } from 'graphql'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../graphql/queries/getApp'
import LOGIN_USER from '../../graphql/queries/loginUser'
import LoginPage from '../../pages/login'
import { useRouter } from 'next/router'

describe('Login Page', () => {
  const fakeUsername = 'fake username'
  const fakePassword = 'fake password'

  const fillOutLoginForm = async getByTestId => {
    const usernameField = getByTestId('username')
    const passwordField = getByTestId('password')

    // the type event needs to be delayed so the Formik validations finish
    await userEvent.type(usernameField, fakeUsername, { delay: 1 })
    await userEvent.type(passwordField, fakePassword, { delay: 1 })
  }
  const { push } = useRouter()
  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('Should redirect to /curriculum on success', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: LOGIN_USER,
          variables: {
            username: fakeUsername,
            password: fakePassword
          }
        },
        result: {
          data: {
            login: {
              success: true,
              username: fakeUsername,
              cliToken: 'fake token',
              error: null
            }
          }
        }
      }
    ]

    const { getByTestId } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <LoginPage />
      </MockedProvider>
    )

    const submitButton = getByTestId('submit')

    await fillOutLoginForm(getByTestId)
    fireEvent.click(submitButton)

    await waitFor(() => expect(push).toBeCalledWith('/curriculum'))
  })

  test('Should set alert visible on invalid credentials', async () => {
    const mocks = [
      {
        request: { query: GET_APP },
        result: {
          data: {
            session: null,
            lessons: [],
            alerts: []
          }
        }
      },
      {
        request: {
          query: LOGIN_USER,
          variables: {
            username: fakeUsername,
            password: fakePassword
          }
        },
        result: {
          errors: [new GraphQLError('UserInputError: User does not exist!')]
        }
      }
    ]

    const { container, getByTestId, findByTestId, findByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <LoginPage />
      </MockedProvider>
    )

    const submitButton = await findByTestId('submit')

    await fillOutLoginForm(getByTestId)
    userEvent.click(submitButton)

    await findByText(mocks[1].result.errors[0].message.split(':')[1].trim())
    expect(container).toMatchSnapshot()
    expect(push).not.toBeCalled()
  })
})
