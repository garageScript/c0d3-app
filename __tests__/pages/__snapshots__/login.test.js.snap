// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Should set alert visible on invalid credentials 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="/docs/learn"
            >
              Learn
            </a>
            <a
              class="nav-item nav-link"
              href="https://discord.gg/c0d3"
              rel="noopener noreferrer"
              target="_blank"
            >
              Community
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
  <div
    class="container-md"
  >
    <div
      class="row mt-5"
    >
      <div
        class="card shadow-sm col-sm-8 col-md-7 col-lg-6 col-xl-6 m-auto px-md-5 border-0"
      >
        <div
          class="card-body text-center pt-5 pb-5"
        >
          <h1
            class="card-title h2 font-weight-bold mb-5"
          >
            Login
          </h1>
          <p
            class="card-text"
          />
          <form
            action="#"
            data-testid="form"
          >
            <div
              class="form-group"
            >
              <div
                class="alert d-flex justify-content-between mt-3 alert-danger"
                role="alert"
              >
                <div>
                  <img
                    class="mr-3 alert-icon"
                    src="/assets/curriculum/icons/exclamation.svg"
                  />
                   User does not exist! 
                </div>
              </div>
              <div
                class="mt-3"
              >
                <input
                  class="form-control form-control-lg font-weight-light mb-3 "
                  data-testid="username"
                  name="username"
                  placeholder="Username"
                  value="fake username"
                />
                <input
                  class="form-control form-control-lg font-weight-light mb-3 "
                  data-testid="password"
                  name="password"
                  placeholder="Password"
                  type="password"
                  value="fake password"
                />
                <button
                  class="btn btn-primary btn-lg btn-block mb-3"
                  data-testid="submit"
                  type="submit"
                >
                  Login to Your Account
                </button>
              </div>
            </div>
          </form>
          <a
            class=""
            href="/forgotpassword"
          >
            Forgot your password?
          </a>
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    Â© Copyright 2021 GarageScript
  </footer>
</div>
`;
